### DISCUSSION TITLE
Custom App build error: field .... will be initialized after field xxx [-Werror, -Wreorder-ctor

### DISCUSSION THREAD

**Discussion by jsevarts** [idaholab/moose#25065] [Q&A Getting Started] [answered]
Hi moose community, I have a custom application I am attempting to compile and keep receiving the following errors and warnings. Eventually, I get `fatal error: too many errors emitted....` Any help would be greatly appreciated.  I am running an Apple M2 Pro, Ventura 13.4.1.

```
.../src/materials/PhysicalMaterial/PhysicalMaterial_Helium.C:159:3: error: field '_ad_T' will be initialized after field '_temp_coeff' [-Werror,-Wreorder-ctor]
  _ad_T((_has_temp && is_ad) ? adCoupledValue("temp") : _ad_zero),
```

I am also getting the following warnings...

`ld: warning: direct access in function 'RayTracingStudy::initialSetup()' from file .../projects/moose/modules/ray_tracing/build/unity_src/.libs/userobjects_Unity.aarch64-apple-darwin20.0.0.opt.o' to global weak symbol 'typeinfo for RayKernelTempl<double>' from file .../projects/moose/modules/ray_tracing/build/unity_src/.libs/raykernels_Unity.aarch64-apple-darwin20.0.0.opt.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.`

I have boiled this warning down to another issue...but I am now sure what this would affect downstream. https://github.com/idaholab/moose/discussions/22804

`In file included from
.../build/unity_src/materials_PhysicalMaterial_Unity.C:5:
.../src/materials/PhysicalMaterial/PhysicalMaterial_Nickel_He.C:381:10: warning: unused variable 'he_specific_heat_alt' [-Wunused-variable]
  double he_specific_heat_alt;`

I tested the app on a linux machine and it compiled fine. 

So far, I have reinstalled mambaforge and rebuilt moose twice with the same result.

Moose was installed successfully and all installation tests passed.

Thanks in advanced for any help!

**Comment by lindsayad** [comment, marked as answer]
You need to ensure that the initialization order in your constructor (.C file) does not go against the order of declarations in your class definition in the .h file. See https://stackoverflow.com/a/4467690/4493669

**Reply by lindsayad** [reply]
I would not worry about the `ld` warnings. With respect to the unused variable warning, I think the warning message is pretty descriptive. `he_specific_heat_alt` is not used in your code. If you delete the `double he_specific_heat_alt` declaration, then that warning should go away.

**Reply by jsevarts** [reply]
@lindsayad ,  thanks for the quick response.  Good to know that the two warnings shouldn't affect anything.

Regarding the initialization order in the constructor file, essentially, everyone else who complies the app on either linux or a mac doesn't see these errors.  Since the app compiles for everyone else I am assuming something is wrong with my setup.  Any ideas?

Similar with the unused variable warning.  Others don't seem to be seeing this warning either.

**Reply by lindsayad** [reply]
It's possible to turn that error into a warning. Perhaps that is the default for other users, although that would be surprising to me. I believe my compiler defaults to treating it as an error. To help you further, you need to paste the constructor of your .C file and the contents of your .h file here

**Reply by jsevarts** [reply]
I ended up going through each error and reordering the .C file to match the .h file, and turned off the warnings with the -Wall flag.  The app built and ran.  Thanks for the help!
